ui/src/app/views/workspace-share/component.tsx

```diff
@@ -0,0 +1,508 @@
+
+});
+
+export const UserRoleOptions = [
+  {value: WorkspaceAccessLevel.READER, label: 'Reader'},
+  {value: WorkspaceAccessLevel.WRITER, label: 'Writer'},
+  {value: WorkspaceAccessLevel.OWNER, label: 'Owner'}
+];
+
+export interface WorkspaceShareState {
+  autocompleteLoading: boolean;
+  autocompleteUsers: User[];
+  userNotFound: string;
+  workspaceShareError: boolean;
+  usersLoading: boolean;
+  userRolesList: UserRole[];
+  workspaceFound: boolean;
+  workspaceUpdateConflictError: boolean;
+  workspace: Workspace;
+  searchTerm: string;
+  dropDown: boolean;
+}
+
+export interface WorkspaceShareProps {
+  workspace: Workspace;
+  accessLevel: WorkspaceAccessLevel;
+  userEmail: string;
+  closeFunction: Function;
+  sharing: boolean;
+}
+
+export class WorkspaceShare extends React.Component<WorkspaceShareProps, WorkspaceShareState> {
+  searchTermChangedEvent: Function;
+  searchingNode: HTMLElement;
+
+  constructor(props: WorkspaceShareProps) {
+    super(props);
+    this.state = {
+      autocompleteLoading: false,
+      autocompleteUsers: [],
+      userNotFound: '',
+      workspaceShareError: false,
+      usersLoading: false,
+      userRolesList: this.props.workspace.userRoles || [],
+      workspaceFound: (this.props.workspace !== null),
+      workspaceUpdateConflictError: false,
+      workspace: this.props.workspace,
+      searchTerm: '',
+      dropDown: false,
+    };
+    this.searchTermChangedEvent = fp.debounce(300, this.userSearch);
+  }
+
+  componentWillMount(): void {
+    document.addEventListener('mousedown', this.handleClickOutsideSearch, false);
+  }
+
+  componentWillUnmount(): void {
+    document.removeEventListener('mousedown', this.handleClickOutsideSearch, false);
+  }
+
+  save(): void {
+    if (this.state.usersLoading) {
+      return;
+    }
+    this.setState({usersLoading: true, workspaceShareError: false});
+    workspacesApi().shareWorkspace(this.state.workspace.namespace,
+      this.state.workspace.id,
+      {workspaceEtag: this.state.workspace.etag, items: this.state.workspace.userRoles})
+      .then((resp: ShareWorkspaceResponse) => {
+        const updatedWorkspace = this.state.workspace;
+        updatedWorkspace.etag = resp.workspaceEtag;
+        updatedWorkspace.userRoles = resp.items;
+        this.setState({usersLoading: false, userNotFound: '',
+          searchTerm: '', workspace: updatedWorkspace});
+        this.props.closeFunction();
+      }).catch(error => {
+        if (error.status === 400) {
+          this.setState({userNotFound: error});
+        } else if (error.status === 409) {
+          this.setState({workspaceUpdateConflictError: true});
+        } else {
+          this.setState({workspaceShareError: true});
+        }
+        this.setState({usersLoading: false});
+      });
+  }
+
+  removeCollaborator(user: UserRole): void {
+    this.setState(({userRolesList, workspace}) => (
+      {userRolesList: fp.remove(({email}) => user.email === email, userRolesList),
```

Comment:
nit: My eyes cannot follow this indentation. I think something like this would be better:

```
... => ({
  userRolesList: ...
  workspace: {
   ...
```